File: androld/app/src/main/java/z3d/filemodel/MeshDataManager.java
package z3d.filemodel;

import android.widget.ListView;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;

import z3d.base.ByteArray;
import z3d.base.CallBackFun;
import z3d.base.MeshData;
import z3d.base.ResGC;
import z3d.res.BaseRes;
import z3d.vo.BindParticle;
import z3d.vo.BoneSocketData;
import z3d.vo.Matrix3D;
import z3d.vo.Quaternion;
import z3d.vo.SkinMesh;
import z3d.vo.Vector2D;

public class MeshDataManager extends ResGC {

    private static MeshDataManager _instance;
    public static MeshDataManager getInstance()  {
        if (MeshDataManager._instance==null) {
            MeshDataManager._instance = new MeshDataManager();
        }
        return MeshDataManager._instance;
    }

    public void getMeshData(String $url, CallBackFun bfun, int batchNum)
    {


        /*
        if (this._dic[$url] && this._dic[$url].ready) {
            $fun(this._dic[$url]);
            this._dic[$url].useNum++;
            return;
        }

        if (this._loadDic[$url]) {
            this._loadDic[$url].push($fun);
            return;
        }

        this._loadDic[$url] = new Array;
        this._loadDic[$url].push($fun);


        ResManager.getInstance().loadRoleRes(Scene_data.fileRoot + $url, ($roleRes: RoleRes) => {
            this.roleResCom($roleRes, $fun);
        }, $batchNum);
        */


    }
    public SkinMesh readData(ByteArray _byte,int $batchNum,String $url,int $version)
    {
        SkinMesh $skinMesh = new SkinMesh();
        $skinMesh.fileScale = _byte.readFloat();
        if ($version >= 19) {
            $skinMesh.tittleHeight = _byte.readFloat();
        } else {
            $skinMesh.tittleHeight = 50;
        }
        $skinMesh.hitBox = new Vector2D(20, 20);
        if ($version >= 23) {
            $skinMesh.hitBox.x = _byte.readFloat();
            $skinMesh.hitBox.y = _byte.readFloat();
        }
        $skinMesh.makeHitBoxItem();

        int meshNum = _byte.readInt();
       HashMap allParticleDic = new HashMap();
        for (int i = 0; i < meshNum; i++) {
            MeshData meshData = new MeshData();

            if ($version >= 21) {
                this.readMesh2OneBuffer(_byte, meshData);
            }
            meshData.treNum = meshData.indexs.size();
            meshData.materialUrl = _byte.readUTF();
            meshData.materialParamData = BaseRes.readMaterialParamData(_byte);
            int particleNum = _byte.readInt();
            for ( int j = 0; j < particleNum; j++) {
                BindParticle bindParticle = new BindParticle(_byte.readUTF(), _byte.readUTF());
                meshData.particleAry.add(bindParticle);
                allParticleDic.put(bindParticle.url,true);
            }

            $skinMesh.addMesh(meshData);

        }

/*
        for (String key in allParticleDic) {
            ParticleManager.getInstance().registerUrl(key);
        }
        */

        $skinMesh.allParticleDic = allParticleDic;

        int bindPosLength = _byte.readInt();

        List<List<Float>> bindPosAry  = new ArrayList<>();
        for (int j = 0; j < bindPosLength; j++) {
            List<Float> ary =   new ArrayList<Float>(Arrays.asList(_byte.readFloat(), _byte.readFloat(), _byte.readFloat(),
                    _byte.readFloat(), _byte.readFloat(), _byte.readFloat()));

            bindPosAry.add(ary);
        }

        this.getBindPosMatrix(bindPosAry, $skinMesh);

        int sokcetLenght    = _byte.readInt();

        $skinMesh.boneSocketDic = new HashMap();

        for (int j = 0; j < sokcetLenght; j++) {
            BoneSocketData boneData = new BoneSocketData();
            boneData.name = _byte.readUTF();
            boneData.boneName = _byte.readUTF();
            boneData.index = _byte.readInt();
            boneData.x = _byte.readFloat();
            boneData.y = _byte.readFloat();
            boneData.z = _byte.readFloat();
            boneData.rotationX = _byte.readFloat();
            boneData.rotationY = _byte.readFloat();
            boneData.rotationZ = _byte.readFloat();

            $skinMesh.boneSocketDic.put(boneData.name, boneData);
        }

        this.dic.put($url,$skinMesh);


        return $skinMesh;
    }

    private void getBindPosMatrix(List<List<Float>> bindPosAry, SkinMesh $skinMesh) {
        List<Matrix3D> ary  = new ArrayList<>();
        List<Matrix3D>  invertAry  =  new ArrayList<>();

        for (int i = 0; i < bindPosAry.size(); i++) {
            List<Float>  objbone = bindPosAry.get(i);

            Quaternion OldQ   = new Quaternion(objbone.get(0), objbone.get(1), objbone.get(2));
            OldQ.setMd5W();
            Matrix3D newM  = OldQ.toMatrix3D();

            newM.appendTranslation(objbone.get(3), objbone.get(4), objbone.get(5));
            invertAry.add(newM.clone());
            newM.invert();

            ary.add(newM);
        }

        $skinMesh.bindPosMatrixAry = ary;
        $skinMesh.bindPosInvertMatrixAry = invertAry;

    }

    public void readMesh2OneBuffer(ByteArray _byte, MeshData meshData) {
        int len = _byte.readInt();

        List<Boolean> typeItem  = new ArrayList<>();
        int dataWidth    = 0;
        for (int i = 0; i < 5; i++) {
            boolean tf = _byte.readBoolean();
            typeItem.add(tf);
            if (tf) {
                if (i == 1) {
                    dataWidth += 2;
                } else {
                    dataWidth += 3;
                }
            }
        }

        dataWidth += 8;

        len *= dataWidth * 4;

        int uvsOffsets = 3; // 1
        int normalsOffsets = uvsOffsets + 2; // 2
        int tangentsOffsets = normalsOffsets + 3; //3
        int bitangentsOffsets = tangentsOffsets + 3; //4
        int boneIDOffsets;
        if (typeItem.get(2)) {//normal
            if (typeItem.get(4)) {
                boneIDOffsets = bitangentsOffsets + 3;
            } else {
                boneIDOffsets = normalsOffsets + 3;
            }
        } else {
            boneIDOffsets = uvsOffsets + 2;
        }
        int boneWeightOffsets = boneIDOffsets + 4;


        byte[] arybuff = new byte[(int)len];
        ByteBuffer data=ByteBuffer.wrap(arybuff);

        meshData.verticeslist= BaseRes.readBytes2ArrayBuffer(_byte, data, 3, 0, dataWidth,0);//vertices
        meshData.uvlist= BaseRes.readBytes2ArrayBuffer(_byte, data, 2, uvsOffsets, dataWidth,0);//uvs
        meshData.normals=  BaseRes.readBytes2ArrayBuffer(_byte, data, 3, normalsOffsets, dataWidth,0);//normals
        meshData.tangents=  BaseRes.readBytes2ArrayBuffer(_byte, data, 3, tangentsOffsets, dataWidth,0);//tangents
        meshData.bitangents=   BaseRes.readBytes2ArrayBuffer(_byte, data, 3, bitangentsOffsets, dataWidth,0);//bitangents
        meshData.boneIDAry=  BaseRes.readBytes2ArrayBuffer(_byte, data, 4, boneIDOffsets, dataWidth, 2);//boneIDAry
        meshData.boneWeightAry=   BaseRes.readBytes2ArrayBuffer(_byte, data, 4, boneWeightOffsets, dataWidth, 1);//boneWeightAry



        meshData.indexs=new ArrayList<>();
        meshData.boneNewIDAry=new ArrayList<>();
        BaseRes.readIntForTwoByte(_byte, meshData.indexs);
        BaseRes.readIntForTwoByte(_byte, meshData.boneNewIDAry);

        meshData.compressBuffer = true;
        meshData.uvsOffsets = uvsOffsets * 4;
        meshData.normalsOffsets = normalsOffsets * 4;
        meshData.tangentsOffsets = tangentsOffsets * 4;
        meshData.bitangentsOffsets = bitangentsOffsets * 4;

        meshData.boneIDOffsets = boneIDOffsets * 4;
        meshData.boneWeightOffsets = boneWeightOffsets * 4;

        meshData.stride = dataWidth * 4;


    }


}

File: listios/StorageStarted/pan3d/filemode/MeshDataManager.m
//
//  MeshDataManager.m
//  iosgl
//
//  Created by zhao on 2/4/2020.
//  Copyright © 2020 zhao. All rights reserved.
//

#import "MeshDataManager.h"
#import "GL_Header.h"
#import "ByteArray.h"
#import "SkinMesh.h"
#import "Scene_data.h"
#import "Vector2D.h"
#import "Quaternion.h"
#import "MeshData.h"
#import "BindParticle.h"
#import "ResManager.h"
#import "BoneSocketData.h"
#import "ParticleManager.h"
@interface MeshDataManager ()
@property(nonatomic,strong)NSMutableDictionary* loadDic;
@end

@implementation MeshDataManager


- (instancetype)init:(Scene3D *)value
{
    self = [super init:value];
    if (self) {
        self.dic=[[NSMutableDictionary alloc]init];
        self.loadDic=[[NSMutableDictionary alloc]init];
        
    }
    return self;
}

-(void)getMeshData:(NSString*)url fun:(SkinMeshBfun)fun batchNum:(int)batchNum;
{
    MeshDataManager* this=self;
    if (this.dic[url] && ((SkinMesh*)this.dic[url]).ready) {
        fun(this.dic[url]);
        return;
    }
    if (this.loadDic[url]) {
        [this.loadDic[url] addObject:fun];
        return;
    }
    this.loadDic[url] =[[NSMutableArray alloc]init];
    [this.loadDic[url] addObject:fun];
    [self.scene3D.resManager loadRoleRes:[[Scene_data default]getWorkUrlByFilePath:url]  fun:^(RoleRes * _Nonnull roleRes) {
        
        [self roleResCom:roleRes fun:^(NSString *localPath) {
            
        }];
    } meshBatchNum:batchNum];
    
}
-(void)roleResCom:(RoleRes*)roleRes fun:(SuccessBlock)fun;
{
    
    MeshDataManager* this=self;
    NSString* url = roleRes.roleUrl;
    SkinMesh* skinMesh = this.dic[url];
    [skinMesh loadMaterial];
    [skinMesh setAction:roleRes.actionAry roleUrl:url];
    
    NSArray* arr=  this.loadDic[url];
    for (int i = 0; i <arr.count ; i++) {
        SkinMeshBfun temp=  arr[i];
        temp(skinMesh);
    }
    skinMesh.ready = YES;
    [this.loadDic removeObjectForKey:url];
}


-(void)readData:(ByteArray*)byte batchNum:(int)batchNum url:(NSString*)url version:(int)version;
{
    SkinMesh* skinMesh  =[[SkinMesh alloc]init:self.scene3D];
    skinMesh.fileScale = [byte readFloat];
    skinMesh.tittleHeight =[byte readFloat];
    skinMesh.hitBox=[[Vector2D alloc]init];
    skinMesh.hitBox.x = [byte readFloat];
    skinMesh.hitBox.y = [byte readFloat];
    
    [skinMesh makeHitBoxItem];
    
    int meshNum = [byte readInt];
    NSMutableDictionary* allParticleDic=[[NSMutableDictionary alloc]init];
    
    for (int i = 0; i < meshNum; i++) {
        
        MeshData* meshData =[[MeshData alloc]init:self.scene3D];
        if (version >= 35) {
            meshData.bindPosAry = [self readBindPosByte:byte];
            [self getBindPosMatrix:meshData.bindPosAry skinMesh:skinMesh];
        }
        [self readMesh2OneBuffer:byte meshData:meshData];
        meshData.trinum =(int) meshData.indexs.count ;
        meshData.materialUrl = [byte readUTF];
        meshData.materialParamData= [BaseRes readMaterialParamData:byte];
        int particleNum = [byte readInt];
        for (int j = 0; j < particleNum; j++) {
            BindParticle* bindParticle=[[BindParticle alloc]init:[byte readUTF] socketName:[byte readUTF]];
            [meshData.particleAry  addObject:bindParticle];
            allParticleDic[bindParticle.url] = [NSNumber numberWithInt:1];
        }
        [skinMesh addMesh:meshData];
        
    }
    for (NSString* key in allParticleDic) {
        //  [[ParticleManager default] registerUrl: allParticleDic[key]];
    }
    skinMesh.allParticleDic = allParticleDic;
    if(version<35){
        int bindPosLength = [byte readInt];
        NSMutableArray<NSArray<NSNumber*>*>* bindPosAry=[[NSMutableArray alloc]init];
        for (int j = 0; j < bindPosLength; j++) {
            NSArray * ary = [[NSArray alloc] initWithObjects:[NSNumber numberWithFloat:[byte readFloat]],[NSNumber numberWithFloat:[byte readFloat]],[NSNumber numberWithFloat:[byte readFloat]],[NSNumber numberWithFloat:[byte readFloat]],[NSNumber numberWithFloat:[byte readFloat]],[NSNumber numberWithFloat:[byte readFloat]], nil];
            [bindPosAry addObject:ary];
        }
        [self getBindPosMatrix:bindPosAry skinMesh:skinMesh];
    }
    
  
    
    int sokcetLenght = [byte readInt];
    
    skinMesh.boneSocketDic = [[NSMutableDictionary alloc]init];
    for (int j = 0; j < sokcetLenght; j++) {
        BoneSocketData* boneData=[[BoneSocketData alloc]init];
        boneData.name = [byte readUTF];
        boneData.boneName =[byte readUTF];
        boneData.index =[byte readInt];
        boneData.x = [byte readFloat];
        boneData.y =[byte readFloat];
        boneData.z = [byte readFloat];
        boneData.rotationX = [byte readFloat];
        boneData.rotationY = [byte readFloat];
        boneData.rotationZ = [byte readFloat];
        skinMesh.boneSocketDic[boneData.name] = boneData;
    }
    self.dic[url] = skinMesh;
    
    
    
}
-( NSMutableArray<NSMutableArray<NSNumber*>*>*)readBindPosByte:(ByteArray*)byte;
{
    int bindPosLength  = [byte readInt];
    NSMutableArray<NSMutableArray<NSNumber*>*>* bindPosAry=[[NSMutableArray alloc]init];
    for (int j   = 0; j < bindPosLength; j++) {
        NSMutableArray<NSNumber*>* ary=[[NSMutableArray alloc]init];
        [ary addObject:[NSNumber numberWithFloat:[byte readFloat]]];
        [ary addObject:[NSNumber numberWithFloat:[byte readFloat]]];
        [ary addObject:[NSNumber numberWithFloat:[byte readFloat]]];
        [ary addObject:[NSNumber numberWithFloat:[byte readFloat]]];
        [ary addObject:[NSNumber numberWithFloat:[byte readFloat]]];
        [ary addObject:[NSNumber numberWithFloat:[byte readFloat]]];
        [bindPosAry addObject:ary];
    }
    return bindPosAry;
    
}
/*
 private readBindPosByte(byte: Pan3dByteArray): Array<Array<number>>  {
 var bindPosLength: number = byte.readInt();
 var bindPosAry: Array<Array<number>> = new Array;
 for (var j: number = 0; j < bindPosLength; j++) {
 var ary: Array<number> = new Array(byte.readFloat(), byte.readFloat(), byte.readFloat(),
 byte.readFloat(), byte.readFloat(), byte.readFloat());
 bindPosAry.push(ary);
 }
 return bindPosAry
 
 }
 */
-(void)getBindPosMatrix:(NSArray<NSArray<NSNumber*>*>*)bindPosAry skinMesh:(SkinMesh*)skinMesh;
{
    NSMutableArray<Matrix3D*>* ary = [[NSMutableArray alloc]init];
    NSMutableArray<Matrix3D*>* invertAry= [[NSMutableArray alloc]init];
    
    for (int i = 0; i < bindPosAry.count; i++) {
        NSArray<NSNumber*>* objbone= bindPosAry[i];
        
        Quaternion* OldQ=[[Quaternion alloc]x:objbone[0].floatValue y:objbone[1].floatValue z:objbone[2].floatValue];
        [OldQ setMd5W];
        Matrix3D* newM   = [OldQ  toMatrix3D];
        [newM appendTranslation:objbone[3].floatValue y:objbone[4].floatValue z:objbone[5].floatValue];
        
        [invertAry addObject:[newM clone]];
        [newM Invert];
        [ary addObject:[newM clone]];
        
    }
    skinMesh.bindPosMatrixAry = ary;
    skinMesh.bindPosInvertMatrixAry = invertAry;
    
    
}

-(void)readMesh2OneBuffer:(ByteArray*)byte meshData:(MeshData*)meshData;
{
    int len    = [byte  readInt];
    NSMutableArray<NSNumber*>* typeItem =[[NSMutableArray alloc]init];
    int dataWidth = 0;
    for (int i = 0; i < 5; i++) {
        BOOL tf   = [byte  readBoolean];
        [typeItem addObject:tf?@1:@0];
        if (tf) {
            if(i == 1){
                dataWidth += 2;
            }else{
                dataWidth += 3;
            }
        }
    }
    dataWidth += 8;
    len *= dataWidth * 4;
    int verOffsets=0;
    int uvsOffsets = 3; // 1
    int normalsOffsets =  uvsOffsets + 2; // 2
    int tangentsOffsets = normalsOffsets + 3; //3
    int bitangentsOffsets = tangentsOffsets + 3; //4
    int boneIDOffsets;
    if([typeItem[2]boolValue]){//normal
        if([typeItem[4]boolValue]){
            boneIDOffsets = bitangentsOffsets + 3;
        }else{
            boneIDOffsets = normalsOffsets + 3;
        }
    }else{
        boneIDOffsets = uvsOffsets + 2;
    }
    int boneWeightOffsets = boneIDOffsets + 4;
    
    int buffStride=dataWidth * 4;
    NSMutableData *dataBase = [[NSMutableData alloc] initWithLength:len];
    
    meshData.vertices=  [BaseRes readBytes2ArrayBuffer:byte nsdata:dataBase dataWidth:3 offset:verOffsets stride:buffStride readType:0];
    meshData.uvs=  [BaseRes readBytes2ArrayBuffer:byte nsdata:dataBase dataWidth:2 offset:uvsOffsets stride:buffStride readType:0];
    meshData.nrms=  [BaseRes readBytes2ArrayBuffer:byte nsdata:dataBase dataWidth:3 offset:normalsOffsets stride:buffStride readType:0];
    meshData.tangents=  [BaseRes readBytes2ArrayBuffer:byte nsdata:dataBase dataWidth:3 offset:tangentsOffsets stride:buffStride readType:0];
    meshData.bitangents=  [BaseRes readBytes2ArrayBuffer:byte nsdata:dataBase dataWidth:3 offset:bitangentsOffsets stride:buffStride readType:0];
    meshData.boneIDAry=  [BaseRes readBytes2ArrayBuffer:byte nsdata:dataBase dataWidth:4 offset:boneIDOffsets stride:buffStride readType:2];
    meshData.boneWeightAry=  [BaseRes readBytes2ArrayBuffer:byte nsdata:dataBase dataWidth:4 offset:boneWeightOffsets stride:buffStride readType:1];
    
    
    
    NSMutableData *indexNsData = [[NSMutableData alloc]init];
    meshData.indexs=   [BaseRes readIntForTwoByte:byte nsdata:indexNsData];
    meshData.boneNewIDAry=   [BaseRes readIntForTwoByte:byte nsdata:indexNsData];
    
    meshData.compressBuffer = YES;
    meshData.uvsOffsets = uvsOffsets * 4;
    meshData.normalsOffsets = normalsOffsets * 4;
    meshData.tangentsOffsets = tangentsOffsets * 4;
    meshData.bitangentsOffsets = bitangentsOffsets * 4;
    meshData.boneIDOffsets = boneIDOffsets * 4;
    meshData.boneWeightOffsets = boneWeightOffsets * 4;
    meshData.stride = dataWidth * 4;
    
    //  [meshData upToGpu];
    
}
-(NSArray<NSNumber*>*)getArrByStr:(NSString*)str;
{
    
    
    
    NSArray<NSString*>* item = [str componentsSeparatedByString:@","]; //分段
    
    NSMutableArray<NSNumber*>* arr=[[NSMutableArray alloc]init];
    for(int i=0;i<item.count;i++){
        
        [arr addObject:   [NSNumber numberWithFloat: [item[i]floatValue]]  ];
        
        //   [arr addObject: @0];
    }
    
    return arr;
}

@end
